compiler/src/emit/jump.rs

```diff
@@ -0,0 +1,96 @@
+use analyzer::types::hir::{Conditional, Loop};
+use analyzer::types::INT;
+use std::mem::size_of;
+
+use crate::bytecode::{Bytecode, Opcode};
+use crate::constant_pool::ConstantPool;
+use crate::emit::{emit, EmissionState};
+
+pub fn emit_conditional(
+    conditional: &Conditional,
+    emitter: &mut Bytecode,
+    cp: &mut ConstantPool,
+    state: &mut EmissionState,
+) {
+    emit(&conditional.condition, emitter, cp, state);
+
+    if conditional.condition.ty == INT {
+        emitter.emit_code(Opcode::ConvertIntToByte);
+        emitter.emit_byte(1);
+        emitter.emit_code(Opcode::BXor);
+    }
+
+    emitter.emit_code(Opcode::IfJump);
+    let jump_to_then_placeholder = emitter.create_placeholder(size_of::<usize>());
+    //jump out of otherwise instructions
+    let mut jump_oo_otherwise_placeholder = None;
+
+    if let Some(otherwise) = &conditional.otherwise {
+        emit(otherwise, emitter, cp, state);
+
+        // then we jump out of the otherwise instruction
+        emitter.emit_code(Opcode::Jump);
+        jump_oo_otherwise_placeholder = Some(emitter.create_placeholder(size_of::<usize>()));
+    }
+
+    emitter.fill_in_ip(jump_to_then_placeholder, emitter.len());
+    emit(&conditional.then, emitter, cp, state);
+    if let Some(jump_placeholder) = jump_oo_otherwise_placeholder {
+        emitter.fill_in_ip(jump_placeholder, emitter.len());
+    }
```

Comment:
There are too many internals used here. Why creating a placeholder here? Why having two `if let Some(_) = otherwise`? This could be rewritten as such:
```rs
    // Evaluate the condition.
    emit(&conditional.condition, emitter, cp, state);
    if conditional.condition.ty == INT {
        emitter.emit_code(Opcode::ConvertIntToByte);
        emitter.emit_byte(1);
        emitter.emit_code(Opcode::BXor);
    }

    // If the condition is false, go to ELSE.
    let jump_to_else = emitter.emit_jump(Opcode::IfNotJump);
    // Evaluate the if branch.
    emit(&conditional.then, emitter, cp, state);
    // Go to END.
    let jump_to_end = emitter.emit_jump(Opcode::Jump);

    // ELSE:
    emitter.patch_jump(jump_to_else);
    if let Some(otherwise) = &conditional.otherwise {
        emit(otherwise, emitter, cp, state);
    }

    // END:
    emitter.patch_jump(jump_to_end);
```

Expose higher level methods in the `Bytecode` struct:
```rs
    /// Emits a jump instruction.
    ///
    /// It returns the index of the offset which is to be patched.
    #[must_use = "the jump address must be patched later"]
    pub fn emit_jump(&mut self, opcode: Opcode) -> usize {
        self.emit_code(opcode);
        self.create_placeholder(size_of::<usize>())
    }

    /// Takes the index of the jump offset to be patched as input, and patches
    /// it to point to the current instruction.
    pub fn patch_jump(&mut self, offset_idx: usize) {
        let ip = self.len();
        self.fill_in_ip(offset_idx, ip);
    }
```